AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -0,0 +1,810 @@
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    captureContext->audioUnit = _audioUnit;
+
+    return YES;
+}
+
+- (BOOL)startAudioUnit {
+    OSStatus status = AudioOutputUnitStart(_audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not start the audio unit!");
+        return NO;
+    }
+    return YES;
+}
+
+- (BOOL)stopAudioUnit {
+    OSStatus status = AudioOutputUnitStop(_audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not stop the audio unit!");
+        return NO;
+    }
+    return YES;
+}
+
+- (void)teardownAudioUnit {
+    if (_audioUnit) {
+        AudioUnitUninitialize(_audioUnit);
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+    }
+}
+
+#pragma mark - NSNotification Observers
+
+- (void)registerAVAudioSessionObservers {
+    // An audio device that interacts with AVAudioSession should handle events like interruptions and route changes.
+    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
+
+    [center addObserver:self selector:@selector(handleAudioInterruption:) name:AVAudioSessionInterruptionNotification object:nil];
+    /*
+     * Interruption handling is different on iOS 9.x. If your application becomes interrupted while it is in the
+     * background then you will not get a corresponding notification when the interruption ends. We workaround this
+     * by handling UIApplicationDidBecomeActiveNotification and treating it as an interruption end.
+     */
+    if (![[NSProcessInfo processInfo] isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){10, 0, 0}]) {
+        [center addObserver:self selector:@selector(handleApplicationDidBecomeActive:) name:UIApplicationDidBecomeActiveNotification object:nil];
+    }
+
+    [center addObserver:self selector:@selector(handleRouteChange:) name:AVAudioSessionRouteChangeNotification object:nil];
+    [center addObserver:self selector:@selector(handleMediaServiceLost:) name:AVAudioSessionMediaServicesWereLostNotification object:nil];
+    [center addObserver:self selector:@selector(handleMediaServiceRestored:) name:AVAudioSessionMediaServicesWereResetNotification object:nil];
+}
+
+- (void)handleAudioInterruption:(NSNotification *)notification {
+    AVAudioSessionInterruptionType type = [notification.userInfo[AVAudioSessionInterruptionTypeKey] unsignedIntegerValue];
+
+    @synchronized(self) {
+        if (self.renderingContext->deviceContext) {
+            TVIAudioDeviceExecuteWorkerBlock(self.renderingContext->deviceContext, ^{
+                if (type == AVAudioSessionInterruptionTypeBegan) {
+                    NSLog(@"Interruption began.");
+                    self.interrupted = YES;
+                    [self stopAudioUnit];
+                } else {
+                    NSLog(@"Interruption ended.");
+                    self.interrupted = NO;
+                    [self startAudioUnit];
```

Comment:
Do we need to do any more to stop AVAudioEngine from attempting to call our stopped audio unit?

Reply:
Tested - AVAudioEngine handles interruption. The music and the remote audio gets stopped when interrupt begins, and music (from where it had stopped) and remote audio resumes on interrupt end.